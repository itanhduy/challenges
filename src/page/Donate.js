import React, { PureComponent } from 'react'
import { Screen, Row, Text, Card, Divider, PaymentOptions } from '../component'
import { API, Transform } from '../service'

class Donate extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      charityInformation: {},
      paymentOptions: {},
      paymentOptionSelected: null,
    }
  }

  componentDidMount() {
    /**
     * Get the topicId from the url
     */
    const { topicId } = this.props.match.params
    /**
     * Start fetching all information
     * First response will be charity information
     * Second response will be payment options for this charity
     */
    Promise.all([API.getCharity(topicId), API.getPayments(topicId)]).then(
      ([charityInfoResponse, paymentOptionsResponse]) => {
        this.setState({
          charityInformation: charityInfoResponse.data,
          paymentOptions: paymentOptionsResponse.data,
        })
      },
    )
  }

  /**
   * Render right component for card to show totalAmount
   * For this charity
   * @return {Text} The text component with totalAmount
   */
  rightComponent = () => {
    const { charityInformation } = this.state
    const { totalAmount, currency } = charityInformation
    return (
      <Text formatMoney={true} currency={currency}>
        {totalAmount}
      </Text>
    )
  }

  /**
   * The function that will be called
   * When user changed payment option
   * @param item The item information of payment option
   */
  optionChange = paymentOptionSelected => {
    this.setState({ paymentOptionSelected })
  }

  /**
   * After user selected payment option
   * Show result so user can know which option they have selected
   * @return {PureComponent} The componen that included option payment result
   */
  renderOptionSelected = () => {
    const { paymentOptionSelected } = this.state
    return (
      paymentOptionSelected && (
        <Text styleName="md-gutter-top medium">You are gonna donate ${paymentOptionSelected.label}</Text>
      )
    )
  }

  /**
   * Render bottom component for this charity
   * @return {PureComponent} The bottom component that included payment options and navigation
   */
  bottomComponent = () => {
    const { paymentOptions, paymentOptionSelected } = this.state
    return (
      <Row styleName="vertical">
        <Divider />
        <Row styleName="vertical md-gutter-top">
          <Text styleName="medium">Payment options</Text>
          <Row styleName="md-gutter-top vertical">
            <PaymentOptions
              data={Transform.createOptionData(paymentOptions, 'amount', ['amount', 'currency'])}
              isGroup={true}
              groupName="payment-options"
              onChange={this.optionChange}
            />
          </Row>
        </Row>
        <Row styleName="md-gutter-top">
          <Divider />
        </Row>
        <Row>{this.renderOptionSelected()}</Row>
      </Row>
    )
  }

  renderView = () => {
    const { charityInformation } = this.state
    return (
      <Screen styleName="h-center xl-gutter">
        <Row styleName="xl-gutter-top">
          <Text styleName="title bold fadeIn">Donation for {charityInformation.name}</Text>
        </Row>
        <Row styleName="xl-gutter-top width-30">
          <Card
            data={charityInformation}
            columns={1}
            rightComponent={this.rightComponent}
            bottomComponent={this.bottomComponent}
            cardContentProps={{ paddingBottom: 0 }}
          />
        </Row>
      </Screen>
    )
  }

  render() {
    return this.renderView()
  }
}

export default Donate
