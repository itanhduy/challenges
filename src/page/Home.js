import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { sumBy, map, cloneDeep } from 'lodash'
import { Responsive } from '../theme'
import { ListCharities, Screen, Text, Row, Button } from '../component'
import { CharityAction, DonationAction } from '../redux/actions'
import { API } from '../service'

class Home extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      charities: [],
      columns: 2,
      widthPercent: 75,
    }
  }

  async componentDidMount() {
    const charitiesResponse = await API.charities()
    const paymentsRespone = await API.payments()
    const newListCharities = this.mapPaymentAmount(charitiesResponse.data, paymentsRespone.data)
    const { updateTotalDonate } = this.props
    this.setState(
      {
        charities: newListCharities,
      },
      () => {
        const totalAmount = sumBy(paymentsRespone.data, item => item.amount)
        updateTotalDonate(totalAmount)
      },
    )
    Responsive.listen([
      {
        minWidth: 0,
        maxWidth: 425,
        inCase: () => {
          this.setState({
            columns: 1,
            widthPercent: 100,
          })
        },
      },
      {
        minWidth: 425,
        maxWidth: 1024,
        inCase: () => {
          this.setState({
            columns: 1,
            widthPercent: 75,
          })
        },
      },
      {
        minWidth: 1024,
        inCase: () => {
          this.setState({
            columns: 2,
          })
        },
      },
    ])
  }

  /**
   * Map totalAmount of charity with payments response
   * @param {Array} listCharities  The list charities
   * @param {Array} payments The list payments
   * @return {Array<Object>} New array of campaign with totalAmount
   */
  mapPaymentAmount = (listCharities, payments) => {
    const newListCharities = map(listCharities, item => {
      const totalAmount = sumBy(payments, payment => {
        return item.id === payment.charitiesId ? payment.amount : 0
      })
      item.totalAmount = totalAmount
      return item
    })
    return newListCharities
  }

  /**
   * When user click on Donate button, run this function
   * @param {Object} item The data of charity
   * @return {Function} Redirect user to donation page
   */
  startDonating = (event, item) => {
    const { history, selectedCampaign } = this.props
    selectedCampaign(item)
    history.push(`donate/${item.id}`)
  }

  /**
   * Return right component for card
   * @param callBackItem The item callback return from the parent component which is wrapping this rightComponent
   * @return {Button} The button element for right position
   */
  rightComponent = callBackItem => {
    return (
      <Button
        textProps={{ styleName: 'textPrimaryColor medium' }}
        onClick={event => this.startDonating(event, callBackItem)}>
        Donate
      </Button>
    )
  }

  /**
   * Return description component for card
   * @param callBackItem The item callback return from the parent component which is wrapping this rightComponent
   * @return {PureComponent} The component for description position
   */
  descriptionComponent = callBackItem => {
    return (
      <Text styleName="caption medium">{`Total raised: ${callBackItem.totalAmount} ${callBackItem.currency}`}</Text>
    )
  }

  render() {
    const { charities, columns, widthPercent } = this.state
    const { header } = this.props
    return (
      <Screen styleName="v-center">
        <Row styleName="xl-gutter-top">
          <Text styleName="title bold fadeIn">{header}</Text>
        </Row>
        <Row styleName={`vertical width-${widthPercent} xl-gutter-top`}>
          <ListCharities
            data={cloneDeep(charities)}
            columns={columns}
            rightComponent={this.rightComponent}
            descriptionComponent={this.descriptionComponent}
          />
        </Row>
      </Screen>
    )
  }
}

Home.propTypes = {
  header: PropTypes.string.isRequired,
}

function mapStateToProps(state) {
  return {
    charity: state.charity,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    /**
     * Dispatch total donate to store
     * @param {Number} total The total amount
     */
    updateTotalDonate: total => {
      dispatch(CharityAction.updateTotalDonate(total))
    },
    /**
     * Dispatch campaign that user is selecting to donate to store
     * @param {Object} campaign The campaign information
     */
    selectedCampaign: campaign => {
      dispatch(DonationAction.selectedCampaign(campaign))
    },
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Home)
