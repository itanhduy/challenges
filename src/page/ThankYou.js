import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { Home } from '../page'
import { Screen, Row, Divider, Text, Nothing, Button, GoBackHome } from '../component'

class ThankYou extends PureComponent {
  /**
   * Check if we don't have enough information
   * Then show Nothing component, otherwise show ThankYou content
   * @return {PureComponent} Nothing component or ThankYou component
   */
  renderView = () => {
    const { donation } = this.props
    const { donationInformation, campaignInformation } = donation.info
    if (!donationInformation && !campaignInformation) {
      return <Nothing />
    }
    return (
      <Screen>
        <Row styleName="lg-gutter-top">
          <Text styleName="bold heading text-center width-100">{`Thank you for donation ${donationInformation.amount} ${
            donationInformation.currency
          } on campaign ${campaignInformation.name}`}</Text>
        </Row>
        <Row styleName="lg-gutter-top">
          <Divider />
        </Row>
        <Home header="You may interested with these campaigns" {...this.props} />

        <Row styleName="lg-gutter flexible v-center">
          <GoBackHome styleName="width-30" />
        </Row>
      </Screen>
    )
  }

  render() {
    return this.renderView()
  }
}

function mapStateToProps(state) {
  return {
    donation: state.donation,
  }
}

export default connect(mapStateToProps)(ThankYou)
