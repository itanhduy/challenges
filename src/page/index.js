import Donate from './Donate'
import Home from './Home'
import ThankYou from './ThankYou'

export { Donate, Home, ThankYou }
