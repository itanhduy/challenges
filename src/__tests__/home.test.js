import React from 'react'
import Store from '../store'
import { Provider } from 'react-redux'
import { Home } from '../page'
import renderer from 'react-test-renderer'

test('Should render Home component', () => {
  const homeComponent = renderer.create(
    <Provider store={Store}>
      <Home header="Omise Tamboon React" />
    </Provider>,
  )
  const tree = homeComponent.toJSON()
  console.info(tree)
  expect(tree).toMatchSnapshot()
})
