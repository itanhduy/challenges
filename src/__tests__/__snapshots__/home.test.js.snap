// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Home component 1`] = `
<div
  className="Screen__ScreenComponent-sc-182dw2i-0 ksPqLC"
>
  <div
    className="Row__RowComponent-sc-6b1omn-0 ilGeGH"
  >
    <div
      className="Text__TextComponent-sc-1bxt3zu-0 iimCvt"
    >
      Omise Tamboon React
    </div>
  </div>
  <div
    className="Row__RowComponent-sc-6b1omn-0 cUkQeO"
  />
</div>
`;
